lib/sla.js

```diff
@@ -189,96 +243,276 @@ function originateChannel(client, inbound, bridge, trunkData) {
           unanswered.channel));
       hangup();
     });
+    updateState(client, bridge.name, 'NOT_INUSE');
     defer.reject(new customError.CustomError('InboundHungup',
           'Inbound channel hungup'));
   }
-
-  trunkData.getAllStations()
-    .then(function (stations) {
-      if (stations.length) {
-        stations.forEach(function(station) {
-          dialed.push({endpoint: station.endpoint, channel: client.Channel()});
-        });
-        dialed.forEach(function (object) {
-          var originate = Q.denodeify(object.channel.originate.bind(
-              object.channel));
-          originate({endpoint: object.endpoint, app: 'sla',
-              appArgs: 'dialed', timeout: 10})
-            .catch(function (err) {
-              defer.reject(err);
+  var defer = Q.defer();
+  var dialed = [];
+  getState(client, bridge.name).then(function(currentState) {
+    if (currentState === 'UNKNOWN') {
+      updateState(client, bridge.name, 'NOT_INUSE');
+      currentState = 'NOT_INUSE';
+    }
+    if (currentState === 'NOT_INUSE') {
+      extension.getAllStations()
+        .then(function (stations) {
+          if (stations.length) {
+            stations.forEach(function(station) {
+              dialed.push({endpoint: station, channel: client.Channel()});
+            });
+            dialed.forEach(function (object) {
+              var originate = Q.denodeify(object.channel.originate.bind(
+                  object.channel));
+              originate({endpoint: object.endpoint, app: 'sla',
+                  appArgs: 'dialed', timeout: 10})
+                .catch(function (err) {
+                  defer.reject(err);
+                });
+              updateState(client, bridge.name, 'RINGING');
+              object.channel.once('StasisStart', dialedEnteredStasis);
+              object.channel.once('ChannelDestroyed', dialedChannelHungup);
             });
-          object.channel.once('StasisStart', dialedEnteredStasis);
-          object.channel.once('ChannelDestroyed', dialedChannelHungup);
+           inbound.once('ChannelHangupRequest', inboundHungup);
+          } else {
+            defer.reject(new customError.CustomError('NoStations',
+                  'No stations in this shared extension.'));
+          }
         });
-        inbound.once('ChannelHangupRequest', inboundHungup);
-      }
-      else {
-        defer.reject(new customError.CustomError('NoStations',
-              'No stations in this trunk.'));
-      }
-    });
+    } else {
+      // This will either redirect this channel towards a Hangup or a "backup
+      // extension"
+      var kickOutUser = Q.denodeify(inbound.continueInDialplan.bind(inbound));
+      kickOutUser();
+      defer.reject(new customError.CustomError('ExtensionOccupied',
+            'An inbound caller attempted to call into a busy extension'));
+    }
+  });
+  return defer.promise;
+}
+
+function originateOutbound(client, outbound, bridge, extension, trunkData) {
```

Comment:
I would reorder the parameters here from most general to specific. client, extension, trunkData, outbound, bridge. Keep things that change often at the end of the function signature.